assets/pages/new-contributions-landing/components/StripePaymentRequestButton.jsx

```diff
@@ -0,0 +1,198 @@
+  const email = data.payerEmail;
+  if (email) {
+    if (isValidEmail(email)) {
+      setEmail(email);
+    } else {
+      logException(`Failed to set email for stripe payment request user with email: ${email}`);
+    }
+  } else {
+    logException('Failed to set email: no email in data object');
+  }
+}
+
+const onComplete = (complete: Function) => (res: PaymentResult) => {
+  if (res.paymentStatus === 'success') {
+    complete('success');
+  } else if (res.paymentStatus === 'failure') {
+    complete('fail');
+  }
+};
+
+function onClick(event, props: PropTypes) {
+  event.preventDefault();
+  props.setStripePaymentRequestButtonClicked();
+  const amountIsValid =
+    checkAmountOrOtherAmount(
+      props.selectedAmounts,
+      props.otherAmounts,
+      props.contributionType,
+      props.countryGroupId,
+    );
+  if (props.paymentRequest && amountIsValid) {
+    props.paymentRequest.show();
+  }
+}
+
+function initialisePaymentRequest(props: PropTypes) {
+  const paymentRequest = props.stripe.paymentRequest({
+    country: props.country,
+    currency: props.currency,
+    total: {
+      label: 'The Guardian',
+      amount: props.amount,
+    },
+    requestPayerEmail: true,
+  });
+  paymentRequest.on('token', ({ complete, token, ...data }) => {
+    // We need to do this so that we can offer marketing permissions on the thank you page
+    updateUserEmail(data, props.updateEmail);
+    const tokenId = props.isTestUser ? 'tok_visa' : token.id;
+    props.onPaymentAuthorised({ paymentMethod: 'Stripe', token: tokenId }, onComplete(complete));
+  });
+
+  // The returned value from canMakePayment will either be:
+  // . null - browser has no compatible payment method)
+  // . {applePay: true} - applePay is available
+  // . {applePay: false} - GooglePay or PaymentRequestApi available
+  paymentRequest.canMakePayment().then((result) => {
+    if (result && result.applePay === true) {
+      props.setCanMakeApplePayPayment(true);
+    }
+  });
+  props.setPaymentRequest(paymentRequest);
+}
+
+const paymentButtonStyle = {
+  paymentRequestButton: {
+    theme: 'dark',
+    height: '42px',
+  },
+};
+
+function updateAmount(amount: number, paymentRequest: Object) {
+  // When the other tab is clicked, the value of amount is NaN
+  if (!Number.isNaN(amount) && paymentRequest) {
+    paymentRequest.update({
+      total: {
+        label: 'The Guardian',
+        amount: amount * 100,
+      },
+    });
+  }
+}
+
+
+// ---- Component ----- //
+function PaymentRequestButton(props: PropTypes) {
+  // If we haven't initialised the payment request, initialise it and return null, as we can't insert the button
+  // until the async canMakePayment() function has been called on the paymentRequest object.
+  if (!props.paymentRequest) {
+    initialisePaymentRequest({ ...props });
+    return null;
+  }
+  if (!props.canMakeApplePayPayment) {
+    return null;
+  }
+
+
+  updateAmount(props.amount, props.paymentRequest);
```

Comment:
I wonder if it's clearer to have this happen in the `onClick`?

Reply:
Yes, great suggestion!